### PR TITLE
feat: Add support for Azure Maps Search APIs

### PR DESCRIPTION
**Pull request by SudhindraKovalam**
- Add support for Azure Maps Get Search API
- Add support for Batch Search API

### CODE REVIEW

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzMapsSearchSchemas.scala [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+package com.microsoft.azure.synapse.ml.geospatial
+
+import com.microsoft.azure.synapse.ml.core.schema.SparkBindings
+import java.util.Date
+
+object Address extends SparkBindings[Address]
+
+case class Address (
+  // The building number on the street. DEPRECATED, use streetNumber instead.
+  buildingNumber: Option[String] = None,
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzMapsSearchSchemas.scala:13]: Can we remove the =None s if they arent used?

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsBase.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.azure.synapse.ml.geospatial
+
+import com.microsoft.azure.synapse.ml.cognitive._
+
+abstract class AzureMapsBase(override val uid: String)
+  extends CognitiveServicesBase(uid) with HasSubscriptionKey with HasInternalJsonOutputParser {
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsBase.scala:9]: These classes seem a bit thin right now, perhaps the traits could be added into your actual models

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      }
+    }
+  }
+
+  protected def getInternalInputParser(schema: StructType): HTTPInputParser = {
+    new CustomInputParser().setNullableUDF(inputFunc(schema))
+  }
+}
+
+trait BatchAsyncReply extends HasAsyncReply with HasSubscriptionKey {
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:108]: Im not seeing much azure maps specific stuff here, is it possible to re-use existing async reply trait to keep it DRY?

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsSearch.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.DataType
+
+object AzureMapsAPIConstants {
+  val DefaultAPIVersion = "1.0"
+}
+
+object SearchAddress extends ComplexParamsReadable[SearchAddress]
+
+class SearchAddress(override val uid: String)
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsSearch.scala:20]: Do we still need this API if we have a higher throughput BatchAPI?

**Review by mhamilton723 — CHANGES_REQUESTED** [review]
Looks really good! A few questions for you

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala (hunk 3 above)
Comment (SudhindraKovalam) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:108]: The primary difference is how auth information is passed. Most Cognitive Services pass in the subscription key as a request header. Also I want to implement response massaging in this trait.

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsBase.scala (hunk 2 above)
Comment (SudhindraKovalam) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsBase.scala]: Done.. Removed these..

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzMapsSearchSchemas.scala (hunk 1 above)
Comment (SudhindraKovalam) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzMapsSearchSchemas.scala]: Done... Cleaned up

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.StructType
+
+import java.net.{URI, URLEncoder}
+import java.util.concurrent.TimeoutException
+import scala.concurrent.blocking
+import scala.language.{existentials, postfixOps}
+import spray.json.DefaultJsonProtocol.{StringJsonFormat, seqFormat}
+
+trait HasAddressInput extends HasServiceParams with HasSubscriptionKey with HasURL {
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:24]: Will this ne needed if theres no more non-batch endpoint

**Review by mhamilton723 — APPROVED** [review]

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala (hunk 5 above)
Comment (SudhindraKovalam) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:24]: Cleaned it up

**Pull request merged**